incubator/virtualcluster/pkg/syncer/syncer.go

```diff
@@ -78,122 +108,164 @@ func New(
-func (s *Syncer) onVirtualClusterUpdate(old, new interface{}) {
-	oldVc, ok := old.(*v1alpha1.Virtualcluster)
-	if !ok {
-		klog.Errorf("cannot convert to *v1alpha1.VirtualCluster: %v", old)
-		return
-	}
-	vc, ok := new.(*v1alpha1.Virtualcluster)
-	if !ok {
-		klog.Errorf("cannot convert to *v1alpha1.VirtualCluster: %v", new)
+func (s *Syncer) removeCluster(key string) {
+	klog.Infof("Remove cluster %s", key)
+	s.Lock()
+	defer s.Unlock()
+
+	if !s.clusterSet.Has(key) {
+		// already deleted
 		return
 	}
 
-	switch {
-	case vc.Status.Phase == v1alpha1.ClusterPending:
-		klog.Infof("Virtualcluster(%s) is pending", vc.Name)
-	case vc.Status.Phase == v1alpha1.ClusterRunning &&
-		oldVc.Status.Phase != v1alpha1.ClusterRunning:
-		if err := s.registerInformerCache(vc); err != nil {
-			klog.Errorf("fail to register informer cache for Virtualcluster(%s): %s", vc.Name, err)
-			return
-		}
-	default:
-		klog.Errorf("unknown Virtualcluster(%s) phase: %s",
-			vc.Name, vc.Status.Phase)
+	innerCluster := &cluster.Cluster{
+		Name: key,
 	}
+	for _, clusterChangeListener := range listener.Listeners {
+		clusterChangeListener.RemoveCluster(innerCluster)
+	}
+
+	s.clusterSet.Delete(key)
 }
 
-func (s *Syncer) onVirtualClusterDelete(obj interface{}) {
-	var vc *v1alpha1.Virtualcluster
-	switch t := obj.(type) {
-	case *v1alpha1.Virtualcluster:
-		vc = t
-	case cache.DeletedFinalStateUnknown:
-		var ok bool
-		vc, ok = t.Obj.(*v1alpha1.Virtualcluster)
-		if !ok {
-			klog.Errorf("cannot convert to *v1.Node: %v", t.Obj)
-			return
-		}
-	default:
-		klog.Errorf("cannot convert to *v1.Node: %v", t)
-		return
+// addCluster registers and start an informer cache for the given VirtualCluster
+func (s *Syncer) addCluster(key string, vc *v1alpha1.Virtualcluster) error {
+	klog.Infof("Add cluster %s", key)
+	s.Lock()
+	if s.clusterSet.Has(key) {
+		return nil
 	}
+	s.Unlock()
 
-	klog.Infof("handle virtual cluster %s delete event", vc.Name)
-
+	adminKubeConfigSecret, err := s.secretClient.Secrets(vc.Namespace).Get(KubeconfigAdmin, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get secret (%s) for virtual cluster %s/%s: %v", KubeconfigAdmin, vc.Namespace, vc.Name, err)
+	}
+	clusterRestConfig, err := clientcmd.RESTConfigFromKubeConfig(adminKubeConfigSecret.Data[KubeconfigAdmin])
+	if err != nil {
+		return fmt.Errorf("failed to build rest config for virtual cluster %s/%s: %v", vc.Namespace, vc.Name, err)
+	}
 	innerCluster := &cluster.Cluster{
-		Name: vc.Name,
+		Name:   conversion.ToClusterKey(vc),
+		Config: clusterRestConfig,
 	}
+
+	// for each resource type of the newly added VirtualCluster, we add a listener
 	for _, clusterChangeListener := range listener.Listeners {
-		clusterChangeListener.RemoveCluster(innerCluster)
+		clusterChangeListener.AddCluster(innerCluster)
 	}
+
+	go func() {
+		if err = innerCluster.Start(s.stopChan); err != nil {
+			s.removeCluster(key)
+		}
+	}()
+
+	s.Lock()
+	s.clusterSet.Insert(key)
```

Comment:
This looks problematic, if inclustercluser.start fails at line 262, should we insert the key? 
We probably should use error chan (check the innersCluster.Start method) to check the go routine error.